### DISCUSSION TITLE
Stripe SDK <> Edge Function Compatibility

### DISCUSSION THREAD

**Discussion by treyhuffine** [vercel/next.js#43976] [Ideas]
### Describe the feature you'd like to request

Stripe updated their SDK so it would work in V8 environments by allowing for the `fetch` API instead of `net/http`. However, it doesn't work in Vercel edge functions because of the Node events library - https://github.com/stripe/stripe-node/blob/dab331e25c8aa1a0c78faff0c4f862f5eb8567b8/src/stripe.ts#L48

It shows the error `Error: The edge runtime does not support Node.js 'events' module.`

I imagine this is one of the most-used packages, and I'm hopeful there's a solution because Cloudflare was able to use the Stripe SDK in their Workers' V8 environment - https://blog.cloudflare.com/announcing-stripe-support-in-workers/

### Describe the solution you'd like

The core functionality of the Stripe SDK doesn't seem to use any native Node modules. Stripe has updated it to function with web APIs. However, it does use Node events, and we would need a way to mock this functionality which I imagine is what others have done to make this work in V8.

### Describe alternatives you've considered

Calling the Stripe API directly instead of using the SDK

**Comment by SukkaW** [comment]
@treyhuffine Would you like to share more details about how Cloudflare Workers was able to run the `stripe` Node.js SDK? IIRC Cloudflare Worker doesn't have Node.js' `events` built-in module as well.

**Comment by treyhuffine** [comment]
Thank you for the quick response @amyegan! Since the architecture for Cloudflare Workers is the same as Vercel Edge Functions, they also have the same restriction of not allowing for Node.js modules, just like Vercel. I know Cloudflare collaborated directly with Stripe to make updates to the Stripe SDK so it would work in the Cloudflare V8 environment, which sounds very similar to the architecture Vercel uses for Edge Function with V8.

From the [Cloudflare article](https://blog.cloudflare.com/announcing-stripe-support-in-workers/):

<img width="788" alt="Screen Shot 2022-12-12 at 5 08 58 PM" src="https://user-images.githubusercontent.com/11709986/207376635-f0f57a80-eea4-4a84-b3c5-965f9f789752.png">

For example, to enable this to work in V8, Stripe introduced the ability in their SDK to inject a `fetch` function so it could pull it from the V8 engine like the browser does and not rely on the Node `http` module. 

```javascript
const Stripe = require("stripe");

const stripe = Stripe(STRIPE_API_KEY, {
  // This uses a native `fetch` now so it can work outside Node
  httpClient: Stripe.createFetchHttpClient()
});
```

[Stripe also added try/catch for native Node requirements](https://github.com/stripe/stripe-node/blob/3e80143129af5046c42cbec338c9049aa525face/src/utils.ts) so it can fail non-essential aspects gracefully since it's not available in the Cloudflare/Vercel V8 architecture.

<img width="613" alt="Screen Shot 2022-12-12 at 5 17 41 PM" src="https://user-images.githubusercontent.com/11709986/207376717-aaf665c1-5f5d-48ad-88e1-82d4d9310619.png">

With payments being such a core part of applications and a majority of apps using Stripe for it, it would be great to cover this and utilize it through Edge Functions since they seem to be much faster than Serverless Functions which require a cold start. The improved speed of Edge Function would provide a better user experience at a key piece of the user’s journey.

Perhaps Vercel has a contact with the Stripe team? I believe they have committed code to Next.js for Stripe implementations.

What can I do to help move this forward?

**Reply by amyegan** [reply]
Thanks for sharing this idea! Since this topic is the same as https://github.com/vercel/community/discussions/1124, let's keep the discussion in one place and continue the conversation there 🙏

**Reply by treyhuffine** [reply]
Thanks @amyegan! I wasn't sure if this was a Vercel or Next challenge to solve, so I posted it in both.
